include/awkward/Index.h

```diff
@@ -90,7 +90,7 @@ namespace awkward {
     IndexOf<T>(const std::shared_ptr<T>& ptr,
                int64_t offset,
                int64_t length,
-               kernel::lib ptr_lib = kernel::lib::cpu);
```

Comment:
It would be fair to say that your troubles with segfaults were mostly due to this line and the equivalent in NumpyArray.h (line 175): letting the default `ptr_lib` label be `kernel::lib::cpu` led to a lot of cases in which a buffer was really on the GPU but the object said it was on the CPU. This happened especially because of `shallow_copy`; these C++ objects are lightweight "handles" to the arrays, so they're shallowly copied all the time, and `shallow_copy` was taking `ptr_lib = kernel::lib::cpu` as a default. The cases that _didn't_ segfault were due to considerable luck that the Index/NumpyArray/Identities weren't copied.

This is why I don't like default parameters (in C++ or in Python) for internal functions. For our purposes, all C++ functions are internal functions, so that's why so little of the C++ codebase has default parameters. Default parameters are great for user convenience, so the highlevel interface has a lot of default parameters, mostly opt-in things with a default of None.

This is similar to using fully qualified names in the code (e.g. `awkward1.operations.convert.from_cupy`) but not in the user interface (e.g. `ak.from_cupy`). For internal functions, convenience is bad and explicitness is good. For users, especially data analysts working interactively in Jupyter, the opposite is true.